mpact_api/mpact/services.py

```diff
@@ -1,49 +1,71 @@
-from telegram.ext import Updater
+from telethon import TelegramClient
 
-from .constants import BOT_TOKEN, ID, MESSAGE, REPLACE, WEBHOOK_URL
-from .models import UserChat, UserData
-from .serializers import ChatDataSerializer, UserDataSerializer
+from .constants import (
+    BOT_TOKEN,
+    CODE,
+    LOGOUT,
+    MESSAGE,
+    MESSAGE_SENT,
+    NOT_AUTHORIZED,
+    PHONE,
+    TELEGRAM_API_HASH,
+    TELEGRAM_API_ID,
+)
+from .logger import logger
 
-updater = Updater(BOT_TOKEN, use_context=True)
-dispatcher = updater.dispatcher
+# Remember to use your own values from my.telegram.org!
+api_id = TELEGRAM_API_ID
+api_hash = TELEGRAM_API_HASH
+bot_token = BOT_TOKEN
+# loop = asyncio.get_event_loop()
+client = TelegramClient("anon", api_id, api_hash)
+bot = TelegramClient("bot", api_id, api_hash).start(bot_token=bot_token)
 
 
-def set_webhook():
-    # TODO: we can move this to config file or keep this as a seperate script.
-    webhook = updater.bot.setWebhook(WEBHOOK_URL, allowed_updates=[MESSAGE])
-    return webhook  # returns True or False
+def connection(func):
+    async def inner_function(*args, **kwargs):
+        try:
+            # Start the client before we start serving
+            await client.connect()
+            results = await func(*args, **kwargs)
+            # After serving (near shutdown), clean up the client
+            await client.disconnect()
+            return results
+        except Exception as exception:
+            logger.exception(exception)
+            raise
 
+    return inner_function
```

Comment:
TelegramClient is already an async context manager. (It extends `telethon.client.auth.AuthMethods`, which implements `__aenter__()` and `__aexit()__` methods, which connect and disconnect for you.) Instead of defining a decorator, you can just use `client` as an async context manager. I'll add an example for `get_dialogs()`.

Reply:
The context manager calls the [start()](https://docs.telethon.dev/en/stable/modules/client.html#telethon.client.auth.AuthMethods.start) by itself and tries to log in if required. So the problem here is in the terminal, it will ask for a phone number and password. So what I did for login is I used client.connect() and for the rest of the APIs, we can use context manager. But if an unauthenticated user tries to access these APIs, it will again ask for a phone number and password in the terminal. So we need to place some condition for this.